plugins/unreal/XJGame/Plugins/XJMusicPlugin/Source/XJMusicPlugin/Private/XjMusicInstanceSubsystem.cpp

```diff
@@ -0,0 +1,188 @@
+// Fill out your copyright notice in the Description page of Project Settings.
+
+
+#include "XjMusicInstanceSubsystem.h"
+#include "Kismet/GameplayStatics.h"
+#include <Settings/XJMusicDefaultSettings.h>
+#include <Sound/SoundBase.h>
+#include <Sound/SoundWave.h>
+#include <Engine/StreamableManager.h>
+#include <Engine/AssetManager.h>
+#include <Engine/ObjectLibrary.h>
+#include <Misc/FileHelper.h>
+#include <Runtime/Engine/Public/AudioDevice.h>
+
+#include "TimerManager.h"
+
+void UXjMusicInstanceSubsystem::RetrieveProjectsContent(const FString& Directory)
+{
+	WorldAudioDeviceHandle = GetWorld()->GetAudioDevice();
+
+	TArray<FString> WavFiles;
+
+	UXJMusicDefaultSettings* XjSettings = GetMutableDefault<UXJMusicDefaultSettings>();
+	if (!XjSettings)
+	{
+		return;
+	}
+
+	FString WorkPath = Directory;
+
+	IPlatformFile& PlatformFile = FPlatformFileManager::Get().GetPlatformFile();
+
+	if (!PlatformFile.DirectoryExists(*WorkPath))
+	{
+		return;
+	}
+
+	TArray<FString> FoundAudioFilesPaths;
+
+	PlatformFile.FindFilesRecursively(FoundAudioFilesPaths, *WorkPath, *AudioExtension);
+
+	for (const FString& Path : FoundAudioFilesPaths)
+	{
+		FString Name = FPaths::GetCleanFilename(Path);
+		AudioPathsByNameLookup.Add(Name, Path);
+	}
+}
+
+void UXjMusicInstanceSubsystem::PlayAudioByName(const FString& Name, const float OverrideStartTime)
+{
+	if (SoundsMap.Contains(Name))
+	{
+		return;
+	}
+
+	USoundWave* SoundWave = GetSoundWaveByName(Name);
+	if (!SoundWave)
+	{
+		return;
+	}
+
+	if (FAudioDeviceHandle Device = GetWorld()->GetAudioDevice())
+	{
+		TSharedPtr<FActiveSound> ActiveSound = MakeShared<FActiveSound>();
+		ActiveSound->SetSound(SoundWave);
+		ActiveSound->SetWorld(GetWorld());
+
+		ActiveSound->SetVolume(1.0f);
+
+		ActiveSound->RequestedStartTime = OverrideStartTime;
```

Comment:
As far as I can tell, the value of OverrideStateTime is always zero- either I'm missing something or there is some missing logic to compute this value

Reply:
It looks like `RequestedStartTime` and `StartTime` should actually be named `SeekTime` https://forums.unrealengine.com/t/audio-component/97378/5

meaning that unfortunately while we are looking for scheduling the start time of the audio against the audio output device, this method is designed for starting playback of the audio in another point than the beginning. Hence default value of 0 will always play the audio from its beginning. Specifing a value here of 0.5 will still play the audio _now_, but it will start 0.5 seconds into the audio.

Reply:
Thats actullay an idea. For example you are on 3.0 second of chain micros and only now you get the audio track that need to be played at 0.0 to 6.0seconds so you now start track starting from it's 3 seconds. Doen't implemented that fully though that's why it's still doesn't compute anything. 
The code where I check for execution time: 
```
for (TimeRecord Time : AudiosKeys)
{
	if (!AudiosLookup.Contains(Time))
	{
		continue;
	}

	TArray<FAudioPlayer>& Audios = AudiosLookup[Time];
	
	Audios.RemoveAll([this, &PlayingAudios](FAudioPlayer& Element)
		{
			if (Element.StartTime > AtChainMicros.GetMicros())
			{
				return false;
			}

			if (Element.EndTime <= AtChainMicros.GetMicros())
			{
				XjMusicInstanceSubsystem->StopAudioByName(Element.Id);
				return true;
			}

			PlayingAudios += FString::Printf(TEXT("%s start - %f  end - %f\n"), *Element.Name, Element.StartTime.GetSeconds(), Element.EndTime.GetSeconds());

			if (Element.bIsPlaying)
			{
				return false;
			}

			Element.bIsPlaying = true;

			XjMusicInstanceSubsystem->PlayAudioByName(Element.Id);
```
As you can see we check if we should play this audio, if not than go to next active audio and if we need to play something we subtract current time from actual time it's had to be started that why I need to get the offset inside audio track